File: src/libtobiid/IDMessage.cpp
/*
    Copyright (C) 2009-2011  EPFL (Ecole Polytechnique Fédérale de Lausanne)
    Michele Tavella <michele.tavella@epfl.ch>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "IDMessage.hpp"
#include <libtobicore/TCException.hpp>
#include <string.h>
#include <stdio.h>

#ifdef __BORLANDC__
using namespace std;
#endif

IDMessage::IDMessage(void) {
	this->Init();
}

IDMessage::IDMessage(IDMessage* const other) {
	this->Init();
	this->Copy(other);
}

IDMessage::IDMessage(IDFtype familyType, IDevent event) {
	this->Init();
	this->_familyType = familyType;
	this->_event = event;
	this->_description = "unset";
}

void IDMessage::Copy(IDMessage* const other) {
	TCBlock::SetBlockIdx(other->GetBlockIdx());
	this->_event = other->GetEvent();
	this->_familyType = other->GetFamilyType();
	this->_description = other->GetDescription();
}

void IDMessage::Init(void) {
	TCBlock::SetBlockIdx(-1);
	this->_familyType = IDMessage::FamilyUndef;
	this->_event = IDMessage::EventNull;
}

IDMessage::~IDMessage(void) {
}

std::string IDMessage::GetDescription(void) const {
	return this->_description;
}

void IDMessage::SetDescription(const std::string& description) {
	this->_description = description;
}

IDFvalue IDMessage::GetFamily(void) const {
	IDFvalue fvalue;
	switch(this->_familyType) {
		case IDMessage::FamilyBiosig:
			fvalue.assign(IDTYPES_FAMILY_BIOSIG);
			break;
		case IDMessage::FamilyUndef:
		default:
			fvalue.assign(IDTYPES_FAMILY_UNDEF);
			break;
	}
	return fvalue;
}

void IDMessage::SetFamilyType(const IDFtype type) {
	this->_familyType = type;
}

IDFtype IDMessage::GetFamilyType(void) const {
	return this->_familyType;
}

void IDMessage::SetEvent(const IDevent event) {
	this->_event = event;
}

IDevent IDMessage::GetEvent(void) const {
	return this->_event;
}

void IDMessage::Dump(void) const {
	printf("[IDMessage::Dump] TOBI iD message for frame %d [%s]\n",
			TCBlock::GetBlockIdx(), this->GetDescription().c_str());
	IDFvalue fvalue = this->GetFamily();
	printf(" + Event family  %d/%s\n", this->GetFamilyType(), fvalue.c_str());
	printf(" + Event value   %d\n", this->GetEvent());

}
		
const IDFtype IDMessage::FamilyType(IDFvalue family) {
	if(family.compare(IDTYPES_FAMILY_BIOSIG) == 0)
		return IDMessage::FamilyBiosig;
	else
		return IDMessage::FamilyUndef;
}

File: doc/examples/mextobiid/idmessage.m
%   Copyright (C) 2010 Michele Tavella <michele.tavella@epfl.ch>
%
%   This file is part of the mexctobiid wrapper
%
%   The libndf library is free software: you can redistribute it and/or
%   modify it under the terms of the version 3 of the GNU General Public
%   License as published by the Free Software Foundation.
%
%   This program is distributed in the hope that it will be useful,
%   but WITHOUT ANY WARRANTY; without even the implied warranty of
%   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%   GNU General Public License for more details.
%
%   You should have received a copy of the GNU General Public License
%   along with this program.  If not, see <http://www.gnu.org/licenses/>.

clear all;

[family_type, family_name] = idmessage_getfamilies();

message1 = idmessage_new();
idmessage_setdescription(message1, 'classifier');
idmessage_setfamilytype(message1, 0);
idmessage_setevent(message1, 781);
idmessage_setbidx(message1, 100);

serializer1 = idserializerrapid_new(message1);
buffer1 = idmessage_serialize(serializer1);
disp(['[example_idmessage] XML message1: ' buffer1]);

message2 = idmessage_new();
idmessage_setdescription(message2, idmessage_getdescription(message1));
serializer2 = idserializerrapid_new(message2);
idmessage_deserialize(serializer2, buffer1);
buffer2 = idmessage_serialize(serializer2);
disp(['[example_idmessage] XML message2: ' buffer2]);

message3 = idmessage_new();
serializer3 = idserializerrapid_new(message3);
idmessage_copy(message3, message1);
buffer3 = idmessage_serialize(serializer3);
disp(['[example_idmessage] XML message3: ' buffer3]);

idmessage_dumpmessage(message1)
disp(['[example_idmessage] Family name: ' ...
	idmessage_getfamily(message1)]);
disp(['[example_idmessage] Family type: '...
	num2str(idmessage_getfamilytype(message1))]);
idmessage_getfamilytype(message1);

client = idasclient_new();
idasclient_setbidx(client, 100);
disp(['[example_idmessage] Client size: ' num2str(idasclient_size(client))]);
idasclient_add(client, message1);
idasclient_add(client, message2);
idasclient_add(client, message3);
disp(['[example_idmessage] Client size: ' num2str(idasclient_size(client))]);


messageT = idmessage_new();
while(true)
	found = idasclient_getthis(client, messageT, [], []);
	idmessage_dumpmessage(messageT);
	if(found == false)
		break;
	end
end
%, idmessage_familytype('biosig'), 781);


idmessage_delete(message1);
idmessage_delete(message2);
idmessage_delete(message3);
idserializerrapid_delete(serializer1);
idserializerrapid_delete(serializer2);
idserializerrapid_delete(serializer3);
idasclient_delete(client);
